File: cp/typescript/sieva.ts
export function sieva(target: number) {
	const primes = new Array(target).fill(true);

	if (target > 0) primes[0] = false;
	if (target > 1) primes[1] = false;

	for (let i = 4; i < target; i += 2) primes[i] = false;

	for (let i = 3; i * i < target; i += 2) {
		if (!primes[i]) continue;

		for (let j = i * i; j < target; j += 2 * i) {
			primes[j] = false;
		}
	}

	return primes;
}

File: cp/cpp/sieva.cpp
#include "sieva.hpp"

using std::vector;

vector<bool> sieva(const std::size_t target)
{
  std::vector<bool> primes(target, true);

  if (target > 0)
  {
    primes[0] = false;
  }
  if (target > 1)
  {
    primes[1] = false;
  }

  for (std::size_t i = 2; i * i < target; ++i)
  {
    if (!primes[i])
    {
      continue;
    }

    for (std::size_t j = i * i; j < target; j += i)
    {
      primes[j] = false;
    }
  }

  return primes;
}
